[win] Fix EH Cont Guard targets when SEH personality is used

dpaoliello opened the pull request:
There were two issues when `/guard:ehcont` is enabled with the SEH personality on Windows:
1. As @namazso correctly identified, we bail out of `WinException::endFunction` early for `MSVC_TableSEH` with funclets, expecting the exception data to be emitted in `endFunclet`, but `endFunclet` didn't copy the EHCont metadata from the function to the module.
2. The SEH personality requires that the basic block containing the `catchpad` is the target, not the `catchret`.

Fixes #64585

dpaoliello committed:
[win] Fix EH Cont Guard targets when SEH personality is used

phoebewang commented:
Thanks a lot for the fixing @dpaoliello !

> In fixing these issues, I renamed the various functions, fields and pass to talk about "EHCont Targets" instead of "EHCont Guardret". This, unfortunately, resulted in a lot of churn for the MIR parser tests.

Would it be better if we split this as an NFC patch?

> 2\. The SEH personality requires that the basic block containing the `catchpad` is the target, not the `catchret`.

Can you provide more details about the reasoning. This was implemented by @arlosi from MS. Although I reviewed it, I don't understand the difference between one way and another.

namazso commented:
@phoebewang that second part addresses the issue I pointed out here: https://github.com/llvm/llvm-project/issues/64585#issuecomment-1672909961

ExceptionHandler and EHcont guard target should always be the same

dpaoliello commented:
> Thanks a lot for the fixing @dpaoliello !
> 
> > In fixing these issues, I renamed the various functions, fields and pass to talk about "EHCont Targets" instead of "EHCont Guardret". This, unfortunately, resulted in a lot of churn for the MIR parser tests.
> 
> Would it be better if we split this as an NFC patch?

Sure, sounds like a good idea.

> > 2. The SEH personality requires that the basic block containing the `catchpad` is the target, not the `catchret`.
> 
> Can you provide more details about the reasoning. This was implemented by @arlosi from MS. Although I reviewed it, I don't understand the difference between one way and another.

I'm aware of that work: Arlo is on my team. EHCont was originally added to LLVM to handle Rust, which always uses the C++ Exception personality (`MSVC_CXX`) and was never tested against SEH. @namazso is correct: in SEH the unwinder will jump to the `ExceptionHandler` address from the unwind table, hence that needs to be marked as safe in the EHCont metadata.

rnk commented:
As high level feedback on the rename, I worry that "ehcont" is opaque jargon. I correctly guessed that it stands for "exception handling continuation", but I don't assume most readers will get that. LLVM definitely doesn't subscribe to the [long descriptive naming philosophy](https://wiki.c2.com/?VeryLongDescriptiveNamesThatProgrammingPairsThinkProvideGoodDescriptions), so I think it's best to split the difference and use the EHCont variable names as you have them, but please update the comments to spell out "EH continuation(s)" to help readers who aren't familiar with Windows EH tech.

In some ways, these points are similar to landing pads, but we've already overloaded "pad" as a concept to include funclet entry blocks. I guess an LLVM "pad" is any basic block that can be the target of a runtime-mediated control flow transfer.

rnk commented:
I think I was in a rush so I forgot to mention, yes, this seems like a good change, thanks for doing the rename to make it consistent. :)

dpaoliello commented:
Split out the renaming as an NFC: #129953

phoebewang reviewed (approved):
The description needs update.
